### PR TITLE
Implement asyncmap/pmap over `map`

### PR DESCRIPTION
**Pull request by amitmurthy**
This is a complete rework of `asyncmap` as a twice `map` on the collection in order to leverage various specializations on `map` for `asyncmap` and `pmap`.

See https://github.com/JuliaLang/julia/issues/19126, https://github.com/JuliaLang/julia/issues/4606 and https://github.com/JuliaLang/julia/issues/14265

```
# currently works as expected for

asyncmap(identity, ones(2,2))
asyncmap(identity, sparse(eye(3)))
asyncmap(identity, (1,2,3))

pmap(identity, ones(2,2))
pmap(identity, sparse(eye(3)))
pmap(identity, (1,2,3))

# WIP for:
asyncmap((x,y,z)->x+y+z, 1,2,3)
asyncmap(x->x, "Hello")
asyncmap(x->x, BitArray(4))
```

This implementation is also a bit faster compared to 0.5
```
julia> @time asyncmap(identity, 1:10^6);
  3.438046 seconds (9.11 M allocations: 264.920 MB, 6.19% gc time)

julia> @time asyncmap(identity, 1:10^6);
  3.399210 seconds (9.00 M allocations: 259.771 MB, 9.18% gc time)
```

0.5:
```
julia> @time asyncmap(identity, 1:10^6);
  3.919530 seconds (34.11 M allocations: 820.279 MB, 25.20% gc time)

julia> @time asyncmap(identity, 1:10^6);
  4.206910 seconds (33.99 M allocations: 815.587 MB, 28.50% gc time)
```

### CODE REVIEW

**Comment by amitmurthy** [comment]
Ready for a review.

File: base/asyncmap.jl [hunk 1]
```diff
@@ -3,161 +3,336 @@
...
+
+If `batch_size` is specified, the collection is processed in batch mode. `f` should
+then be a function that should accept a `Vector` of argument tuples and should
+return a vector of results. The input vector will have a length upto `batch_size`
+
+The following examples return the `object_id` of the tasks in which the mapping
+function is executed for different invocations of `ntasks` and `batch_size`.
+
+With ntasks undefined, each element is processed in a different task.
+```
```
Comment (tkelman) [inline base/asyncmap.jl]: should this be a doctest?

Comment (amitmurthy) [inline base/asyncmap.jl]: I considered that. Since the examples are printing out `object_id`s, wouldn't the doctest fail?

**Comment by kshyatt** [comment]
What's the status on this?

**Comment by amitmurthy** [comment]
Ready to be merged. Just leaving it open for a couple more days for feedback.

**Comment by shashi** [comment]
> See #19126, #4606 and #4606

Did you mean to link to a third issue?

**Comment by amitmurthy** [comment]
Yes - https://github.com/JuliaLang/julia/issues/14265

File: base/asyncmap.jl [hunk 2]
```diff
@@ -3,161 +3,336 @@
...
-Output is collected into `results`.
+    asyncmap(f, c...; ntasks=0, batch_size=nothing)
+
+Uses multiple concurrent tasks to map `f` over a collection (or multiple
+equal length collections). For multiple collection arguments, `f` is
+applied elementwise.
+
+`ntasks` specifies the number of tasks to run concurrently.
+Depending on the length of the collections, if `ntasks` is unspecified,
+upto a 100 tasks will be used for concurrent mapping.
```
Comment (tkelman) [inline base/asyncmap.jl]: "up to" is 2 words

File: base/asyncmap.jl [hunk 3]
```diff
@@ -3,161 +3,336 @@
...
+ 0x440f8819a1baa682
+ 0x9fb3eeadd0c83985
+ 0xebd3e35fe90d4050
+ 0x29efc93edce2b961
+
+julia> length(unique(asyncmap(x->tskoid(), 1:5)))
+5
+```
+
+With `ntasks=2` all elements are only processed in 2 tasks.
```
Comment (tkelman) [inline base/asyncmap.jl]: "only" is unneeded here

File: base/asyncmap.jl [hunk 4]
```diff
@@ -3,161 +3,336 @@
...
+ 0x027ab1680df7ae94
+ 0xa23d2f80cd7cf157
+ 0x027ab1680df7ae94
+
+julia> length(unique(asyncmap(x->tskoid(), 1:5; ntasks=2)))
+2
+```
+
+With `batch_size` defined, the mapping function needs to be changed to accept an array
+of argument tuples. And return an array of results. In this example we call `map`
```
Comment (tkelman) [inline base/asyncmap.jl]: "tuples, and return" otherwise this isn't a sentence

File: base/asyncmap.jl [hunk 5]
```diff
@@ -3,161 +3,336 @@
...
+on the batch input to achieve the same.
```
Comment (tkelman) [inline base/asyncmap.jl]: not at all clear what "the same" is referring to

File: base/asyncmap.jl [hunk 6]
```diff
@@ -3,161 +3,336 @@
...
-        max_tasks = ()->ntasks
-    elseif isa(ntasks, Integer)
-        max_tasks = ()->ntasks
-    elseif isa(ntasks, Function)
-        max_tasks = ntasks
+function async_usemap(f, c...; ntasks=0, batch_size=nothing)
+    ntasks = verify_ntasks(c[1], ntasks)
+
+    if isa(batch_size, Number)
+        @assert batch_size >= 0
```
Comment (tkelman) [inline base/asyncmap.jl]: should give an actual error message here

File: base/asyncmap.jl [hunk 7]
```diff
@@ -3,161 +3,336 @@
...
 end
 
+function wrap_n_exec_twice(chnl, worker_tasks, ntasks, exec_func, c...)
+    # The driver task, creates a Ref object and writes it and the args tuple to
+    # the communication channel for processing by a free worker task.
+    push_arg_to_channel = (x...) -> (r=Ref{Any}(nothing); put!(chnl,(r,x));r)
+
+    if isa(ntasks, Function)
+        map_f = (x...) -> begin
+            # check number of tasks everytime, and start one if required.
```
Comment (tkelman) [inline base/asyncmap.jl]: "every time" is 2 words

File: base/asyncmap.jl [hunk 8]
```diff
@@ -3,161 +3,336 @@
...
-        try
-            for (i, args) in itr.task_chnl
-                state.nfree -= 1
+    # check and throw any exceptions from the worker tasks
+    foreach(x->(v=wait(x); isa(v, Exception) && throw(v)), worker_tasks)
 
-                itr.results[i] = itr.f(args...)
-                notify(state.item_done, nothing)
+    # check if there was a genuine problem with asyncrun
+    (asyncrun_excp != nothing) && throw(asyncrun_excp)
```
Comment (tkelman) [inline base/asyncmap.jl]: use `===` and `!==` to compare to `nothing`

File: base/asyncmap.jl [hunk 9]
```diff
@@ -3,161 +3,336 @@
...
-            # sync_end() then re-throws "e" in the main task.
-            state.in_error = true
-            clear_collector_channel(itr)
-            notify(state.item_done, nothing)
+    if isa(asyncrun, Ref)
+        # scalar case
+        return asyncrun.x
+    else
+        # second run, extract values from the Refs and return
+        return map(x->x.x, asyncrun)
```
Comment (tkelman) [inline base/asyncmap.jl]: use a different variable name when the field is `.x` ?

File: base/asyncmap.jl [hunk 10]
```diff
@@ -3,161 +3,336 @@
...
+            nt=1
         end
+    else
+        nt = ntasks
     end
 
-    state.active_count += 1
-    t
+    # Use an unbuffered channel for communicating with the worker tasks. In the event
+    # of any error in any of the worker tasks, the channel is closed. And this
```
Comment (tkelman) [inline base/asyncmap.jl]: "And" should either be continuing the previous sentence or removed

File: base/asyncmap.jl [hunk 11]
```diff
@@ -3,161 +3,336 @@
...
+function asyncmap(f, s::AbstractString; ntasks=0, batch_size=nothing)
+    s2=Array(Char, length(s))
+    asyncmap!(f, s2, s)
+    return convert(String, s2)
+end
 
-    for _ in 1:itr.max_tasks()
-        start_collector_task(itr, state)
-        state.nfree += 1
+# map on a single BitArray returns a BitArray of the mapping function is boolean.
```
Comment (tkelman) [inline base/asyncmap.jl]: if the mapping function

File: base/asyncmap.jl [hunk 12]
```diff
@@ -3,161 +3,336 @@
...
-        state.done = true
-        close(itr.task_chnl)
-        sync_end()
-    end
-    return state.done
+Returns an iterator which spplies `f` to each element of `c` and
+collects output into results.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
Comment (tkelman) [inline base/asyncmap.jl]: missing open [

File: base/asyncmap.jl [hunk 13]
```diff
@@ -3,161 +3,336 @@
...
+be a function which operates on an array of argument tuples.
+
+!!! note
+    `next(::AsyncCollector, state) -> (nothing, state)`
```
Comment (tkelman) [inline base/asyncmap.jl]: explain why this is significant

File: base/asyncmap.jl [hunk 14]
```diff
@@ -3,161 +3,336 @@
...
+
+function done(itr::AsyncCollector, state::AsyncCollectorState)
+    if !isopen(state.chnl) || done(itr.enumerator, state.enum_state)
+        close(state.chnl)
+
+        # wait for all tasks to finish
+        foreach(x->(v=wait(x); isa(v, Exception) && throw(v)), state.worker_tasks)
+        empty!(state.worker_tasks)
+        return true
+     else
```
Comment (tkelman) [inline base/asyncmap.jl]: indentation is off

File: base/asyncmap.jl [hunk 15]
```diff
@@ -3,161 +3,336 @@
...
-    AsyncGenerator(f, c...; ntasks=0) -> iterator
+    AsyncGenerator(f, c...; ntasks=0, batch_size=nothing) -> iterator
 
 Apply `f` to each element of `c` using at most `ntasks` asynchronous tasks.
-If `ntasks` is unspecified, uses `max(100, nworkers())` tasks.
-For multiple collection arguments, apply `f` elementwise.
-Results are returned by the iterator as they become available.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
Comment (tkelman) [inline base/asyncmap.jl]: missing [

File: base/asyncmap.jl [hunk 16]
```diff
@@ -220,26 +384,22 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
...
-"""
-    asyncmap!(f, c)
+    asyncmap!(f, c; ntasks=0, batch_size=nothing)
 
 In-place version of [`asyncmap()`](:func:`asyncmap`).
 """
-asyncmap!(f, c) = (for x in AsyncCollector(f, c, c) end; c)
-
+function asyncmap!(f, c; ntasks=0, batch_size=nothing)
+    foreach(_->_, AsyncCollector(f, c, c; ntasks=ntasks, batch_size=batch_size))
```
Comment (tkelman) [inline base/asyncmap.jl]: identity is easier to read than `_->_`

**Comment by amitmurthy** [comment]
Incorporated feedback. Meanwhile sparse had been updated with a specialized implementation for `map` which was causing this PR to fail. For now, this PR has a naive implementation - should be optimized later. Added a couple more tests.

File: base/asyncmap.jl [hunk 17]
```diff
@@ -3,161 +3,365 @@
...
-    end
-    return state.done
+"""
+    AsyncCollector(f, results, c...; ntasks=0, batch_size=nothing) -> iterator
+
+Returns an iterator which applies `f` to each element of `c` asynchronously
+and collects output into results.
+
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+`[asyncmap()`](:func:`asyncmap`). If `batch_size` is specified, `f` must
```
Comment (tkelman) [inline base/asyncmap.jl]: I think it's supposed to be backticks inside square brackets

File: base/asyncmap.jl (hunk 14 above)
Comment (tkelman) [inline base/asyncmap.jl]: not addressed

**Comment by amitmurthy** [comment]
Rebased after https://github.com/JuliaLang/julia/pull/18588 . `make docs` is not capturing updated docstrings. Anything else to be done?

File: base/asyncmap.jl [hunk 18]
```diff
@@ -3,161 +3,365 @@
...
-Output is collected into `results`.
+    asyncmap(f, c...; ntasks=0, batch_size=nothing)
+
+Uses multiple concurrent tasks to map `f` over a collection (or multiple
+equal length collections). For multiple collection arguments, `f` is
+applied elementwise.
+
+`ntasks` specifies the number of tasks to run concurrently.
+Depending on the length of the collections, if `ntasks` is unspecified,
+up to a 100 tasks will be used for concurrent mapping.
```
Comment (tkelman) [inline base/asyncmap.jl]: up to 100

File: base/asyncmap.jl [hunk 19]
```diff
@@ -3,161 +3,365 @@
...
-        @assert false "Error should have been captured and thrown previously."
-    end
+    AsyncCollector(f, r, en::Enumerate, ntasks, batch_size) = new(f, r, en, ntasks, batch_size, isa(ntasks, Function))
+end
+
+"""
+    AsyncCollector(f, results, c...; ntasks=0, batch_size=nothing) -> iterator
+
+Returns an iterator which applies `f` to each element of `c` asynchronously
+and collects output into results.
```
Comment (tkelman) [inline base/asyncmap.jl]: code highlight `results`

File: base/asyncmap.jl [hunk 20]
```diff
@@ -3,161 +3,365 @@
...
-        close(itr.task_chnl)
-        sync_end()
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+[`asyncmap()`](@ref). If `batch_size` is specified, `f` must
+be a function which operates on an array of argument tuples.
+
+!!! note
+    `next(::AsyncCollector, state) -> (nothing, state)`. A successful return
+    from `next` indicates that the next element from the input collection is
+    being processed asynchronously. It blocks till a free worker task becomes
```
Comment (tkelman) [inline base/asyncmap.jl]: blocks until

File: base/asyncmap.jl [hunk 21]
```diff
@@ -3,161 +3,365 @@
...
+    available.
+
+!!! note
+    `for _ in AsyncCollector(f, results, c...) end` is equivalent to
+    `map!(f, results, c...)`.
```
Comment (tkelman) [inline base/asyncmap.jl]: is it equivalent to map! or asyncmap! ?

File: base/asyncmap.jl [hunk 22]
```diff
@@ -220,26 +413,21 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
...
+function asyncmap!(f, c; ntasks=0, batch_size=nothing)
+    foreach(identity, AsyncCollector(f, c, c; ntasks=ntasks, batch_size=batch_size))
+    c
+end
 
 """
-    asyncmap!(f, results, c...)
+    asyncmap!(f, results, c...; ntasks=0, batch_size=nothing)
 
 Like [`asyncmap()`](@ref), but stores output in `results` rather returning a collection.
```
Comment (tkelman) [inline base/asyncmap.jl]: rather than returning

File: base/asyncmap.jl (hunk 21 above)
Comment (amitmurthy) [inline base/asyncmap.jl]: Both. The former in terms of returned data. The latter in terms of asynchronous execution.

File: base/asyncmap.jl (hunk 20 above)
Comment (amitmurthy) [inline base/asyncmap.jl]: Indian English comes naturally to me. Will update.

File: base/asyncmap.jl (hunk 21 above)
Comment (amitmurthy) [inline base/asyncmap.jl]: Will update.

**Comment by MichaelHatherly** [comment]
> Rebased after #18588 . `make docs` is not capturing updated docstrings. Anything else to be done?

Docs for `asyncmap` and `asyncmap!` appear to be showing up when I build this branch locally. Are there any error messages being printed?

**Comment by amitmurthy** [comment]
Shows up in the REPL. Doesn't show up in `doc/_build/html/stdlib/parallel.html`. 
```
Documenter: setting up build directory.
Documenter: expanding markdown templates.
Documenter: building cross-references.
Documenter: running document checks.
 !! Skipped doctesting.
 > checking footnote links.
 !! Unused footnote named 'BH83'. [src/stdlib/linalg.md]
Documenter: populating indices.
Documenter: rendering document.
 !! Overwriting '_build/html/en/assets/documenter.css'.
 !! Overwriting '_build/html/en/assets/documenter.js'.
 !! Overwriting '_build/html/en/assets/search.js'.
Build finished. The HTML pages are in _build/html.
```

**Comment by MichaelHatherly** [comment]
> `doc/_build/html/stdlib/parallel.html`

Should be `doc/_build/html/en/stdlib/parallel.html` instead. Perhaps that's just an older build in `doc/_build/html/stdlib/parallel.html`.

**Comment by amitmurthy** [comment]
Right! Thanks.

Merging this PR in a short while.

File: base/asyncmap.jl [hunk 23]
```diff
@@ -220,26 +413,22 @@ size(itr::AsyncGenerator) = size(itr.collector.enumerator)
...
+    asyncmap!(f, results, c...; ntasks=0, batch_size=nothing)
 
-Like [`asyncmap()`](@ref), but stores output in `results` rather returning a collection.
+Like [`asyncmap()`](@ref), but stores output in `results` rather than
+returning a collection.
 """
-asyncmap!(f, r, c1, c...) = (for x in AsyncCollector(f, r, c1, c...) end; r)
+function asyncmap!(f, r, c1, c...; ntasks=0, batch_size=nothing)
+    foreach(identity, AsyncCollector(f, r, c1, c...; ntasks=ntasks, batch_size=batch_size))
+    c
```
Comment (andreasnoack) [inline base/asyncmap.jl]: @amitmurthy This should `r`, right?

Comment (amitmurthy) [reply base/asyncmap.jl]: Yes.

**Pull request merged**